Jordan struct

GDeFranceschi opened the pull request:
This branch is rebased on the branch `VectSpaces` and contains two new files: `src/Groups/matrices/linear_conjugate.jl` and `src/Groups/matrices/linear_centralizer.jl`. Some methods from Hecke are imported to compute the Jordan form (see file `matrix_manipulation.jl`).
The exported methods do the following:
- compute the centralizer of every element in `GL(n,q)` and `SL(n,q)` and compute its cardinality.
- given `G= GL(n,q)` or `SL(n,q)` and given two elements `x`, `y` in `G`, establish whether `x` and `y` are conjugate in `G` and, if yes, returns `z` such that `x^z = y`.
- some further method is exported : `issemisimple`, `isunipotent`, `multiplicative_jordan_decomposition`, `pol_elementary_divisors`, that can be useful in future.

Possible issues: the methods are implemented to bypass the GAP methods, hence the GAP underlying objects of `G`, `x` and `y` are still not defined after calling `centralizer` or `isconjugate` (unless they were already defined, of course). Unlike other types of groups, the line `centralizer(G,x)` returns `(C,Nothing)` (where `C` is the centralizer) instead of `(C,f)` where `f` is the embedding homomorphism of `C` into `G`: that's because, at the moment, homomorphisms between groups (including `MatrixGroups`) are only defined by an underlying GAP object.

ThomasBreuer commented on src/Groups/directproducts.jl:
```diff
@@ -213,21 +193,21 @@ function _as_subgroup(G::DirectProductGroup, H::GapObj)
+  H = GAP.Globals.Subgroup(G.X,elems_in_GAP)
   #H is the group. I need to return the inclusion map too
   return _as_subgroup(G, H)
 end
```
I think it would be simpler to define
```
function sub(G::T, elms::Vector{<:GAPGroupElem{T}}) where T<:GAPGroup
  elems_in_GAP = GAP.julia_to_gap(GapObj[x.X for x in elms])
  H = GAP.Globals.Subgroup(G.X,elems_in_GAP)
  #H is the group. I need to return the inclusion map too
  return _as_subgroup(G, H)
end
```
This would cover also the `sub` methods for semidirect products and wreath products.
(But why is this method needed at all, in addition to the one in `sub.jl`?)

GDeFranceschi replied:
Indeed, I have just checked that for products of groups (direct, semidirect, wreath), I need to redefine only the method `_as_subgroup_bare` (because the constructors are different). The rest is not necessary and I can remove it.

ThomasBreuer commented on src/Groups/homomorphisms.jl:
```diff
@@ -28,9 +28,9 @@ export
-  println(io, codomain(x))
+  show(IOContext(io, :compact => true), domain(x))
+  print(io, "\nto\n")
+  show(IOContext(io, :compact => true), codomain(x))
```
Currently there seem to be no `show` methods for the groups in question that honor the `:compact` flag. Is this something for later?

GDeFranceschi replied:
Matrix groups use the `:compact` flag. This is because the name of a matrix group could be quite long, so to print a homomorphism it seems better to print just the name of the variable identifying the domain/codomain whenever this is a matrix group. But, of course, we can change idea and leave the name of the group. Or, do the opposite and extend it to the other types of groups.

ThomasBreuer commented on src/Groups/homomorphisms.jl:
```diff
@@ -386,20 +386,21 @@ function hom(x::GAPGroupElem{AutomorphismGroup{T}}) where T <: GAPGroup
   G = parent(x)
   if check
-    @assert A.G == G || GAP.Globals.IN(x.X, A.G.X) "Not in the domain of f!"
+    @assert A.G == G || GAP.Globals.IN(x.X, A.G.X) "Not in the domain of f!"      #TODO Do we really need the IN check?
```
Concerning the question:
Without the `IN` check, GAP either shows an error message that is not helpful (here we can argue that GAP should give a better error message) OR returns a group element (this happens for example if the map `f` is a conjugator automorphism on permutations, which can be applied to any permutation in GAP).
If we want to force that `f` can be applied only to elements of its domain then the check is necessary.

GDeFranceschi replied:
The idea is to allow to compute homomorphisms only on elements of the domain, even if it makes sense for larger sets (e.g. conjugator automorphisms defined only on proper subgroups). The `check` variable was not intended to be used by the upper level user, instead by the programmer who writes an algorithm involving automorphisms. If he wants to apply automorphisms on elements that are sure to belong to the domain, he can deactivate the `check` in order to avoid the `IN` check, that could be a quite expensive procedure.

ThomasBreuer commented on src/Groups/matrices/iso_oscar_gap.jl:
```diff
@@ -0,0 +1,118 @@
+   f_gap = GAP.Globals.UnivariatePolynomial(GAP.Globals.GF(p),L_gap)
+   F_gap = GAP.Globals.GF(GAP.Globals.GF(p),f_gap)
+   Basis_F = GAP.Globals.Basis(F_gap)
+   homom(x::fq_nmod) = elem_f(x,GAP.gap_to_julia(GAP.Globals.BasisVectors(Basis_F)),d)
```
Perhaps compute `GAP.gap_to_julia(GAP.Globals.BasisVectors(Basis_F))` once in `gen_ring_iso`, then it need not be computed each time when `homom` gets called.

GDeFranceschi replied:
Good spot. I did the same with the variable `myOne`, but I missed this.

ThomasBreuer commented on src/Groups/matrices/iso_oscar_gap.jl:
```diff
@@ -0,0 +1,118 @@
+function mat_gap_oscar(x::GapObj, riso)
+   n = GAP.Globals.NrRows(x)
+   Arr = [GAP.gap_to_julia(x[i]) for i in 1:n]
+   L = [riso(Arr[i][j]) for i in 1:n for j in 1:n]
```
Here it is assumed that `x` is a list of lists, and that its rows can be accessed. This is the case for the currently used GAP matrices, but does not hold in general.
A general and also simpler alternative is `L = [riso(x[i, j]) for i in 1:n for j in 1:n]`.

ThomasBreuer commented on src/Groups/matrices/iso_oscar_gap.jl:
```diff
@@ -0,0 +1,118 @@
+   end
+
+   return GAP.julia_to_gap(S)
+end
```
This function returns one possible matrix object, but perhaps there are better ones.
For example, the elements of GAP's `GL(n,q)` (for small `q`) are "compressed" matrices.
I am not sure what would be best to achieve nicer matrices here. Either we have to provide a function in GAP.jl that applies some heuristics, or the object `riso` must know about the intended GAP type of the result.

GDeFranceschi replied:
I just used the standard GAP constructor for matrices. Anyway, the operations involving GAP procedures should deal only with elements of `GL(n,q)`, so if it's possible to write a "compressed" matrix, that would be fine.

ThomasBreuer replied:
I would write `return GAP.Globals.ImmutableMatrix( F, GAP.julia_to_gap(S), true)` where `F` is the GAP field from `riso` hich contains the matrix entries. (Note that this field may be larger than the field generated by the matrix entries.)

GDeFranceschi replied:
Done it.

fingolfin replied:
There are many further optimizations possible here, but they require deeper level access to Nemo/Flint matrices; I'll add some suggestions in an issue later

ThomasBreuer replied:
Yes.
My idea behind proposing the `ImmutableMatrix` call here was that the GAP object obtained this way is **what** one wants on the GAP side: a matrix that knows the context `F` (if possible). Improvements **how** one gets such a GAP object are of course welcome.

ThomasBreuer commented on src/Groups/matrices/linear_centralizer.jl:
```diff
@@ -0,0 +1,395 @@
+      for i in 1:n-1 h2[i+1,i] = -1 end
+      return h1,h2
+   end
+end
```
These are the generators from [the paper by Taylor](https://www.gap-system.org/Manuals/doc/ref/chapBib.html#biBTay87), which are used also in GAP. In would propose to give a reference to this paper.

fingolfin replied:
Indeed, I think almost every single function in this file should have a comment at the start giving a reference where one can lookup why the code below might be right -- even to those which compute "basic" things like the orders of GL_n, SL_n, but most definitely anything that's more advanced than that (i.e. not taught in an undergrad course).

Commit:
GAP: deal with matrices, vectors, finite field elements

Commit:
GAP: compute Jordan forms, centralizers,

This branch is rebased on the branch `VectSpaces` and contains two new files:
`src/Groups/matrices/linear_conjugate.jl` and
`src/Groups/matrices/linear_centralizer.jl`. Some methods from Hecke are
imported to compute the Jordan form (see file `matrix_manipulation.jl`). The
exported methods do the following:

- compute the centralizer of every element in `GL(n,q)` and `SL(n,q)` and
  compute its cardinality.
- given `G= GL(n,q)` or `SL(n,q)` and given two elements `x`, `y` in `G`,
  establish whether `x` and `y` are conjugate in `G` and, if yes, returns `z`
  such that `x^z = y`.
- some further method is exported : `issemisimple`, `isunipotent`,
  `multiplicative_jordan_decomposition`, `pol_elementary_divisors`, that can be
 useful in future.

Possible issues: the methods are implemented to bypass the GAP methods, hence
the GAP underlying objects of `G`, `x` and `y` are still not defined after
calling `centralizer` or `isconjugate` (unless they were already defined, of
course). Unlike other types of groups, the line `centralizer(G,x)` returns
`(C,Nothing)` (where `C` is the centralizer) instead of `(C,f)` where `f` is
the embedding homomorphism of `C` into `G`: that's because, at the moment,
homomorphisms between groups (including `MatrixGroups`) are only defined by an
underlying GAP object.

fingolfin commented on src/Groups/matrices/iso_oscar_gap.jl:
```diff
@@ -89,7 +89,7 @@ function mat_oscar_gap(x::fq_nmod_mat, riso)
    end
 
-   return GAP.Globals.ImmutableMatrix( riso.codomain, GAP.julia_to_gap(S), true)
+   return GAP.julia_to_gap(S)
```
I think this is an accidental change?

GDeFranceschi replied:
If we decide to maintain `GAP.Globals.ImmutableMatrix`, yes. I'll restore it.

fingolfin replied:
Please restore it

fingolfin commented on src/Groups/matrices/linear_centralizer.jl:
```diff
@@ -0,0 +1,400 @@
+   res = q^div(n*(n-1),2)
+   for i in 2:n res *= (q^i-1) end
+   return res
+end
```
Lots of untested code in this file...

fingolfin commented on src/Groups/matrices/linear_centralizer.jl:
```diff
@@ -0,0 +1,400 @@
+      h2[1,n] = 1
+      for i in 1:n-1 h2[i+1,i] = -1 end
+      return h1,h2
+   end
```
```suggestion
   n != 1 || return [matrix(F,1,1,[primitive_element(F)])]
   h1 = identity_matrix(F,n)
   h2 = zero_matrix(F,n,n)
   if order(F) == 2
      h1[1,2] = 1
   else
      h1[1,1] = primitive_element(F)
      h2[1,1] = -1
   end
   h2[1,n] = 1
   for i in 1:n-1 h2[i+1,i] = -1 end
   return h1, h2
```

fingolfin commented on src/Groups/matrices/linear_centralizer.jl:
```diff
@@ -0,0 +1,400 @@
+# assumes V is sorted (e.g. [1,1,1,2,3,3])
+function _centr_unipotent(F::Ring, V::AbstractVector{Int}; isSL=false)
+   n = sum(V)
+   _lambda = gen(F)  # yes, gen(F) is correct; we don't need a primitive element in this case
```
What *is* `gen(F)`, actually, for a general ring? Are you perhaps assuming F is a finite field? OK, then what is it for a finite field if not a primitive element?

GDeFranceschi replied:
According to the documentation:
>  gen(a::FqFiniteField)

>  Return the generator of the finite field. Note that this is only guaranteed to be a multiplicative generator if
  the finite field is generated by a Conway polynomial automatically.

In this case I need just `gen(F)`, i.e. a generator of `F` as a field. When I need a generator of `F*` as cyclic group, I use `primitive_element(F)`.

ThomasBreuer replied:
Since we had this discussion also for GAP:
The notation in the literature is not uniform.
In some books (and also in GAP), the term "primitive element" is used for any element that generates the given field (over the given subfield), and any generator of the multiplicative group of a field is called "primitive root".

GDeFranceschi replied:
I can rename `primitive_element` to `primitive_root`.

ThomasBreuer replied:
> I can rename `primitive_element` to `primitive_root`.

I think this will be helpful.
As far as I see, `gen` means "primitive element" in the sense of a generator of a field extension:
The docstring for `gen(L::SimpleNumField)` mentions explicitly that the result is "the canonical primitive element of L over K".

fingolfin commented on src/Groups/matrices/linear_centralizer.jl:
```diff
@@ -0,0 +1,400 @@
+            z = insert_block(identity_matrix(F,n),z,pos,pos)
+            push!(listgens,z)
+         end
+         end
```
You can combine nested for loops:
```suggestion
         for i in 1:l[1]-1, j in 1:degree(F)
            z = identity_matrix(F,l[1])
            for k in 1:l[1]-i z[k,i+k]=_lambda^j end
            z = insert_block(identity_matrix(F,n),z,pos,pos)
            push!(listgens,z)
         end
```

fingolfin commented on src/Groups/matrices/linear_centralizer.jl:
```diff
@@ -0,0 +1,400 @@
+   res = prod([_GL_order(l[2],F) for l in L])
+   exp = fmpz(0)
+   for i in 1:length(L)-1
+   for j in i+1:length(L)
```
merge the nested loops?

GDeFranceschi replied:
Done it.

fingolfin reviewed (commented):
A few comments that apply to many places in here -- please make sure to address them everywhere, not just in the spots where I marked them.

GDeFranceschi committed:
Suggestions from PR#268 + more tests

fingolfin commented on src/Groups/matrices/linear_centralizer.jl:
```diff
@@ -0,0 +1,385 @@
+# returns as matrices
+# does the same as above with F = F[x]/(f), but every entry is replaced by a diagonal join of D corresponding blocks
+# ASSUMPTION: deg(f) > 1
+function _gens_for_GL_matrix(f::PolyElem, n::Int, F::Ring; D=1)
```
```suggestion
function _gens_for_GL_matrix(f::PolyElem, n::Int, F::FinField; D::Int=1)
```

fingolfin commented on src/Groups/matrices/linear_centralizer.jl:
```diff
@@ -0,0 +1,385 @@
+# generators for GL(n,q) and SL(n,q) are described in:  Taylor, D. E., Pairs of Generators for Matrix Groups. I, The Cayley Bulletin, 3 (1987)
+
+# returns as matrices
+function _gens_for_GL(n::Int, F::Ring)
```
This is only valid for F a finite field, right? So perhaps like this?
```suggestion
function _gens_for_GL(n::Int, F::FinField)
```

GDeFranceschi replied:
Yes, better to restrict to finite fields.

fingolfin commented on src/Groups/matrices/linear_centralizer.jl:
```diff
@@ -0,0 +1,385 @@
+end
+
+# returns as matrices
+# does the same as above with F = F[x]/(f), but every entry is replaced by a diagonal join of D corresponding blocks
```
I don't understand. So the `f` is so that you can work with a specific extension of `F`. 

And the D just is used to blow up the matrix? That seems a bit unusual -- you probably have a specific application for it, I'd mention it, and perhaps also say "for details, refer to page X in my thesis" or so (assuming one can find that there)

GDeFranceschi replied:
Yes, `f` is an irreducible polynomial. Since it is one of the output of the function `generalized_jordan_form`, I did not put a check to assure `f` is irreducible.
If we want to find a specific reference on what the function `_gens_for_GL_matrix` does, we can say it gives the generators for `R`, where `centralizer(G,x) = semidirect_product(U,R)`, (see page 22 of article ...). Should I mention it?

fingolfin commented on src/Groups/matrices/linear_centralizer.jl:
```diff
@@ -0,0 +1,385 @@
+   f=ED[1][1]
+   V=[ED[1][2]]
+   c = evaluate(_centralizer(f), companion_matrix(f))
+   c = c^(_disc_log(det(c),_lambda))  # TODO this _disc_log is bad. Don't try with large fields.
```
So what does this do... it compute `d` such that `c = lambda^d`, then compute `c^d = lambda^{d^2}`, is that right?

GDeFranceschi replied:
No. It computes `d` such that `det(c)^d = lambda` and replace `c` by `c^d`. The result `c^d` is a primitive root for the base ring of `x` such that `det(c^d) = lambda`.

fingolfin replied:
Can you please add this explanation in a comment?

GDeFranceschi replied:
Just did it.

fingolfin commented on src/Groups/matrices/linear_centralizer.jl:
```diff
@@ -0,0 +1,385 @@
+end
+
+
+# generators for centralizer of unipotent elements are described in: Giovanni De Franceschi, Centralizers and conjugacy classes in finite classical groups, arXiv:2008.12651
```
Please avoid super wide lines (I usually break after about 80-100 characters), they are difficult to read in e.g. a web browser (as I am doing right now). So e.g.
```suggestion
# generators for centralizer of unipotent elements are described in:
# Giovanni De Franceschi, Centralizers and conjugacy classes in finite classical groups, arXiv:2008.12651
```

Also: could you be a bit more specific, and e.g. say in which section or theorem or whatever one should look? Even if you think it should be "obvious", I think this can be a serious time safer.

fingolfin commented on src/Groups/matrices/linear_centralizer.jl:
```diff
@@ -0,0 +1,385 @@
+# V = vector of integers of the dimensions of Jordan blocks
+# return the generators for the centralizers of the diagonal join of unipotent Jordan blocks of dimensions V
+# assumes V is sorted (e.g. [1,1,1,2,3,3])
+function _centr_unipotent(F::Ring, V::AbstractVector{Int}; isSL=false)
```
So based on what you write elsewhere, perhaps this change should be made:
```suggestion
function _centr_unipotent(F::FinField, V::AbstractVector{Int}; isSL=false)
```
And similarly elsewhere?

Commit:
Fix ImmutableMatrix in iso_oscar_gap.jl

fingolfin commented on src/Groups/matrices/linear_centralizer.jl:
```diff
@@ -0,0 +1,385 @@
+      v_g = isSL ? _gens_for_SL(l[2],F) : _gens_for_GL(l[2],F)
+      for x in v_g
+         z = block_matrix(l[2],l[2],[x[i,j]*identity_matrix(F,l[1]) for i in 1:l[2] for j in 1:l[2]])
+         z = insert_block(identity_matrix(F,n),z,pos,pos)
```
You keep generating fresh identity matrices, use them, then discard them. It may be better to generate them once as early as possible and reuse them as long as possible. E.g. like this (but similar changes should be applied elsewhere in this file):
```suggestion
   idN = identity_matrix(F,n)
   for l in L
      v_g = isSL ? _gens_for_SL(l[2],F) : _gens_for_GL(l[2],F)
      idL1 = identity_matrix(F,l[1])
      for x in v_g
         z = block_matrix(l[2],l[2],[x[i,j]*idL1 for i in 1:l[2] for j in 1:l[2]])
         z = insert_block(idN,z,pos,pos)
```

fingolfin commented on src/Groups/matrices/linear_centralizer.jl:
```diff
@@ -0,0 +1,385 @@
+      pos += L[i][1]*L[i][2]
+      # block above diagonal
+      z = identity_matrix(F,n)
+      for j in 1:L[i][1] z[pos-L[i][1]+j-1,pos+L[i+1][1]-L[i][1]+j-1]=1 end
```
This is really difficult to read. I'd not squish it into a single line. Also, perhaps use more bespoke names?

Maybe this (please double check I didn't mess it up)
```suggestion
      for j in 1:L[i][1]
         idx = pos-L[i][1]+j-1
         z[idx,idx+L[i+1][1]] = 1
      end
```
or
```suggestion
      idx = pos-L[i][1]-1
      for j in 1:L[i][1]
         idx += 1
         z[idx,idx+L[i+1][1]] = 1
      end
```

GDeFranceschi replied:
Corrected in all the places it occurs.

fingolfin commented on src/Groups/matrices/linear_centralizer.jl:
```diff
@@ -0,0 +1,385 @@
+      push!(listgens,z)
+      # block below diagonal
+      z = identity_matrix(F,n)
+      for j in 1:L[i][1] z[pos+j-1,pos-L[i][1]+j-1]=1 end
```
same as above

GDeFranceschi committed:
Suggestions from PR #268

GDeFranceschi committed:
One more tests + comments

Commit:
Add a comment about the discrete logarithm

GDeFranceschi committed:
Replace primitive_root by Hecke.primitive_element

fingolfin commented on src/Groups/matrices/linear_centralizer.jl:
```diff
@@ -0,0 +1,426 @@
+   end
+   h2[1,n] = 1
+   for i in 1:n-1 h2[i+1,i] = -1 end
+   return h1,h2
```
I believe this returns a tuple, while in case `n=1` you return a `Vector` with one element; this violates type stability. I would suggest to just always return a vector? I.e.,
```suggestion
   return [h1,h2]
```

GDeFranceschi replied:
Done it.

fingolfin commented on src/Groups/matrices/linear_centralizer.jl:
```diff
@@ -0,0 +1,426 @@
+   return h1,h2
+end
+
+# returns as matrices
```
```suggestion
```
I find this confusing; after all, both functions return matrices.

GDeFranceschi replied:
It was just to point out that the output elements have type `MatElem`, and not `MatrixGroupElem` (that is something one could expect when computing a centralizer in a group). Should I just remove that line? Or rewrite it?

fingolfin replied:
rewrite it to say what you just did?

GDeFranceschi replied:
I wrote something like `# returns elements of type MatElem{T}`

fingolfin commented on src/Groups/matrices/linear_centralizer.jl:
```diff
@@ -0,0 +1,426 @@
+# does the same as above with the following changes:
+# the field F is replaced by F = F[x]/(f);
+# every entry y is replaced by a diagonal join of D copies of phi(y),
+# where phi is the ring automorphism sending a fixed root of f into the companion matrix of f.
```
Wait, if phi sends roots of f (in F, I assume) to a matrix, it can't be an automorphism?!?
```suggestion
# where phi is the ring homomorphism sending a fixed root of f into the companion matrix of f.
```
Perhaps better to use explicit signatures, i.e. say
> where `phi: F \to ???`
and specify the image ring (not sure what it is?!?)

GDeFranceschi replied:
I don't know why I wrote "automorphism", of course it is not. I just wrote better comment lines.

fingolfin commented on src/Groups/matrices/linear_centralizer.jl:
```diff
@@ -0,0 +1,426 @@
+# the field F is replaced by F = F[x]/(f);
+# every entry y is replaced by a diagonal join of D copies of phi(y),
+# where phi is the ring automorphism sending a fixed root of f into the companion matrix of f.
+# (hence, the final output are still matrices with base ring F)
```
But instead of degree n, those matrices now have degree... uh... `n*deg(f)*D` ? Or something like that? Perhaps this could be mentioned here, to help people check their mental model of this.

fingolfin commented on src/Groups/matrices/linear_centralizer.jl:
```diff
@@ -0,0 +1,426 @@
+   insert_block!(h1,diagonal_join([CP for i in 1:D]),1,1)
+   h2 = zero_matrix(F,n*degree(f)*D,n*degree(f)*D)
+   for i in 1:(n-1)*degree(f)*D h2[i+degree(f)*D,i]=-1 end
+   for i in 1:degree(f)*D
```
This function uses `degree(f)*D` *a lot*; hopefully the Julia compiler will figure it is constant and reuse the value... But I think it would actually help readability if you defined once at the start something like `D2 = degree(f)*D` and used that (or call it `dD` or `E` or whatever you like)

fingolfin commented on src/Groups/matrices/linear_centralizer.jl:
```diff
@@ -0,0 +1,426 @@
+   h1 = identity_matrix(F,n*degree(f)*D)
+   insert_block!(h1,diagonal_join([CP for i in 1:D]),1,1)
+   h2 = zero_matrix(F,n*degree(f)*D,n*degree(f)*D)
+   for i in 1:(n-1)*degree(f)*D h2[i+degree(f)*D,i]=-1 end
```
Please avoid one line loops like this and turn this into three lines

fingolfin commented on src/Groups/matrices/linear_centralizer.jl:
```diff
@@ -0,0 +1,426 @@
+      h2[1,1] = -1
+   end
+   h2[1,n] = 1
+   for i in 1:n-1 h2[i+1,i] = -1 end
```
Please turn into three lines

fingolfin commented on src/Groups/matrices/linear_centralizer.jl:
```diff
@@ -0,0 +1,426 @@
+# assumes V is sorted (e.g. [1,1,1,2,3,3])
+function _centr_unipotent(F::FinField, V::AbstractVector{Int}; isSL=false)
+   n = sum(V)
+   _lambda = gen(F)  # yes, gen(F) is correct; we don't need a primitive root in this case
```
Just for my understanding: So what is `gen(F)`, then?!? Perhaps an element such that its powers form a basis of `F` over the prime field / base field?

I hink you need at least *something* for _lambda, otherwise you could use `one(F)` or even `zero(F)` here; might be a good idea to state in a code comment what it is you do need, not just what you do *not* need... :-)

GDeFranceschi replied:
`gen(a::FqFiniteField)` :  Return the generator of the finite field.
So, `gen(F)` generates `F` as a ring, namely every element of `F` can be written as sum of powers of `gen(F)`. So, `one(F)` is not good whenever `degree(F)` is greater than 1. The difference with `Hecke.primitive_element(F)` is that `gen(F)` does not need to be a generator for the multiplicative group of `F`. I expect that computing `gen(F)` is much simpler than computing `Hecke.primitive_element(F)`, so in this case I used `gen(F)`, because that's all I need.

fingolfin commented on src/Groups/matrices/linear_centralizer.jl:
```diff
@@ -0,0 +1,426 @@
+
+# V = vector of integers of the dimensions of Jordan blocks
+# assumes V is sorted (e.g. [1,1,1,2,3,3])
+# does the same as above, but every entry is replaced by the corresponding block matrix in the field F/(f)
```
Hrm, it's a bit frustrating to have all this code duplication. Surely there must be a way to avoid this... Perhaps not in this PR, but at some point...

GDeFranceschi replied:
The fact is that the first function works with single entries, the second one works with matrix blocks. But I'll see if I can improve it.

fingolfin commented on src/Groups/matrices/linear_centralizer.jl:
```diff
@@ -0,0 +1,426 @@
+   return listgens, res
+end
+
+# returns the list of generators
```
It also returns two more values `res` and `cbm`, can you please also describe them in this comment, even if briefly?

GDeFranceschi replied:
Done it. I just moved the change of basis from the function `centralizer` to the function `_centralizer_GL` and `_centralizer_SL`, since I just realized that it is a cleaner way to proceed.

fingolfin commented on src/Groups/matrices/linear_centralizer.jl:
```diff
@@ -0,0 +1,426 @@
+end
+
+# generators for centralizer of unipotent elements are described in:
+# Giovanni De Franceschi, Centralizers and conjugacy classes in finite classical groups, arXiv:2008.12651
```
Please add a bibtex entry for this to `docs/oscar_references.bib`.

fingolfin commented on src/Groups/matrices/linear_centralizer.jl:
```diff
@@ -0,0 +1,426 @@
+   CP = evaluate(_centralizer(f),C)            # matrix of maximal order in the centralizer of the companion matrix
+   CPi = inv(CP)
+
+   n != 1 || return []
```
Please make sure to preserve type stability; at the very least, this should be
```suggestion
   n != 1 || return MatElem[]
```

GDeFranceschi replied:
Done it.

fingolfin commented on src/Groups/matrices/linear_centralizer.jl:
```diff
@@ -0,0 +1,426 @@
+# returns as matrices
+# returns a generating set for SL(n,F) of at most two elements
+function _gens_for_SL(n::Int, F::FinField)
+   n != 1 || return []
```
```suggestion
   n != 1 || return MatElem[]
```

fingolfin commented on src/Groups/matrices/linear_centralizer.jl:
```diff
@@ -0,0 +1,426 @@
+   end
+   h2[1,n] = 1
+   for i in 1:n-1 h2[i+1,i] = -1 end
+   return h1,h2
```
```suggestion
   return [h1,h2]
```

fingolfin commented on src/Groups/matrices/linear_centralizer.jl:
```diff
@@ -0,0 +1,426 @@
+#  because h1,h2 generate just the subgroup of SL(n*deg(f),F) isomorphic to SL(n, F^deg(f))
+   h3 = identity_matrix(F,n*degree(f)*D)
+   insert_block!(h3,diagonal_join([CP^(order(F)-1) for i in 1:D]),1,1)
+   return h1,h2,h3
```
```suggestion
   return [h1,h2,h3]
```

fingolfin commented on src/Groups/matrices/linear_centralizer.jl:
```diff
@@ -0,0 +1,426 @@
+      h2[i,i]=-1
+      h2[i,i+(n-1)*degree(f)*D]=1
+   end
+# TODO waiting for a better solution, the h3 generator is necessary
```
It is unclear to me what or whom this waiting for... Is this a research question? Then perhaps arXiv:2008.12651 discusses it, and we could point to that discussion here.

Or is this about some function or code missing? Then say what is missing, and how it should be used here once it's there.

Or is it something else? Again, please elaborate

GDeFranceschi replied:
If `H` is an intermediate subgroup between `GL` and `SL`, we know how to generate `H` using three generators, and that's what I do in this piece of code. I'm much confident (if not sure) that it is possible to generate such `H` with only two generators, but I do not know how to do it at the moment. If in future we know how to generate such `H` for every finite field and dimension, the code can be improved by producing a smaller number of generator.
But it's an improvement from 3 to 2; if it is not worth it, I can just delete that line that could create confusion.

fingolfin replied:
So, if you included the explanation you just gave me here into that comment, that'd make me happy :-)

fingolfin commented on src/Groups/matrices/linear_centralizer.jl:
```diff
@@ -0,0 +1,426 @@
+   V=[ED[1][2]]
+   c = evaluate(_centralizer(f), companion_matrix(f))
+# Computes d such that det(c)^d = lambda and replace c by c^d.
+# The result c^d is a primitive root for the base ring of x such that det(c^d) = lambda
```
```suggestion
   # Computes d such that det(c)^d = lambda and replace c by c^d.
   # The result c^d is a primitive root for the base ring of x such that det(c^d) = lambda
```

fingolfin commented on src/Groups/matrices/linear_centralizer.jl:
```diff
@@ -0,0 +1,426 @@
+# The result c^d is a primitive root for the base ring of x such that det(c^d) = lambda
+   c = c^(_disc_log(det(c),_lambda))
+# block_dim is a list of [d,m,c], where
+# d = dimension of the Jordan block, m = its multiplicity, c = el of max order determinant in centralizer of the companion matrix
```
```suggestion
   # block_dim is a list of triples [d,m,c], where
   # d = dimension of the Jordan block, m = its multiplicity, c = el of max order determinant in centralizer of the companion matrix
```

fingolfin commented on src/Groups/matrices/linear_centralizer.jl:
```diff
@@ -0,0 +1,426 @@
+"""
+    centralizer(G::MatrixGroup{T}, x::MatrixGroupElem{T})
+Return (`C`,`Nothing`), where `C` is the centralizer of `x` in `C`.
+To get the embedding homomorphism of `C` into `G`, type
```
```suggestion
To get the embedding homomorphism of `C` into `G`, use
```

GDeFranceschi replied:
Yeah, much better.

fingolfin commented on src/Groups/matrices/linear_centralizer.jl:
```diff
@@ -0,0 +1,426 @@
+
+"""
+    centralizer(G::MatrixGroup{T}, x::MatrixGroupElem{T})
+Return (`C`,`Nothing`), where `C` is the centralizer of `x` in `C`.
```
```suggestion

Return (`C`,`nothing`), where `C` is the centralizer of `x` in `C`.
```
But that's not quite accurate, it only returns `nothing` if `G` is GL or SL.

GDeFranceschi replied:
Just corrected it.

fingolfin commented on src/Groups/matrices/linear_isconjugate.jl:
```diff
@@ -0,0 +1,155 @@
+
+"""
+    multiplicative_jordan_decomposition(M::MatrixGroupElem)
+Return `S` and `U` in the group `G = parent(M)` such that `S` is semisimple, `U` is unipotent and  `M = SU = US`.
```
Please add these empty lines everywhere:
```suggestion

Return `S` and `U` in the group `G = parent(M)` such that `S` is semisimple, `U` is unipotent and  `M = SU = US`.
```

fingolfin commented on src/Groups/matrices/linear_isconjugate.jl:
```diff
@@ -0,0 +1,155 @@
+    multiplicative_jordan_decomposition(M::MatrixGroupElem)
+Return `S` and `U` in the group `G = parent(M)` such that `S` is semisimple, `U` is unipotent and  `M = SU = US`.
+!!! warning "WARNING:" 
+    this is *NOT*, in general, the same output returned when `M` has type `MatElem`.
```
perhaps elaborate as to why that is, and/or even include a concrete example?

fingolfin commented on src/Groups/matrices/linear_isconjugate.jl:
```diff
@@ -0,0 +1,155 @@
+
+"""
+    isunipotent(x::MatrixGroupElem)
+Return whether `x` is unipotent, i.e. its order is a power of the characteristic of its base ring.
```
Those are not the general descriptions of being semisimple or unipotent; these hold at most over finite fields (obviously). 

So perhaps the method signature can be restricted to `MatrixGroupElem{T} where T<:FinFieldElem` or something like that?

GDeFranceschi replied:
Done it.
Though, I think the concept of semisimple and unipotent matrix works for matrices over infinite fields too. The definition is not given in the same terms, of course, since the order may be infinite.

fingolfin replied:
The concepts of unipotent and semisimple elements of course exist over infinite fields (I'd argue they come from there), but the definition you give does not work there; rather, I'd usually give your definition as the result of a lemma that characterize such elements over finite fields. Hence the request to restrict the method signature -- we can provide more generic methods for other fields later, of course

fingolfin commented on src/Groups/matrices/linear_isconjugate.jl:
```diff
@@ -0,0 +1,155 @@
+
+# first: brute force way
+# x has type MatrixGroupElem
+function _elem_given_det(x,d)
```
"brute force way" of what? does that refer to the function `_elem_given_det`? What does it do?!

GDeFranceschi replied:
Yes. To find an element in the centralizer of `x`, it computes the generators of the centralizer and multiplies them in order to obtain an element with the appropriate determinant, that is not the cleanest way to proceed. Do I specify this in the comment?

fingolfin replied:
Yes, all this extra information would be useful in a comment

GDeFranceschi replied:
Added it.

Commit:
Suggestions from #PR 268

GDeFranceschi committed:
Rewrote 4 comments in linear_centralizer.jl

GDeFranceschi committed:
Fixed some comments from PR #268

fingolfin commented on src/Groups/matrices/linear_centralizer.jl:
```diff
@@ -0,0 +1,449 @@
+   c = evaluate(_centralizer(f), companion_matrix(f))
+   # Computes d such that det(c)^d = lambda and replace c by c^d.
+   # The result c^d is a primitive root for the base ring of x such that det(c^d) = lambda
+   c = c^(_disc_log(det(c),_lambda))
```
CI tests fail... around this line?

GDeFranceschi replied:
Solved.

GDeFranceschi committed:
Remove evaluate(::Pol, ::Mat)
